neonvm/runner/main.go

```diff
@@ -779,6 +786,47 @@ func listenForCPUChanges(ctx context.Context, logger *zap.Logger, port int32, cg
 	}
 }
 
+// writes from socket line by line
+func forwardLogs(ctx context.Context, logger *zap.Logger, wg *sync.WaitGroup) error {
+	defer wg.Done()
+
+	delay := 3 * time.Second
+
+	var conn net.Conn
+	var reader *bufio.Reader
+
+	for {
+		if conn == nil {
+			var err error
+			conn, err = net.Dial("unix", logSerialSocket)
+			if err != nil {
+				logger.Error("failed to dial to logSerialSocket", zap.Error(err))
+				time.Sleep(delay)
+				continue
+			}
+			reader = bufio.NewReader(conn)
+		}
+
+		select {
+		case <-ctx.Done():
+			conn.Close()
+			return ctx.Err()
+		default:
+			conn.SetReadDeadline(time.Now().Add(delay))
+			str, err := reader.ReadString('\n')
```

Comment:
a couple things:

1. It's probably cheaper to keep the output as `[]byte` rather than converting to `string` and printing (in general: if the data is already `[]byte`, then strings are less efficient because they're guaranteed read-only)
2. AFAICT there's no length limit here — in theory, someone could send an arbitrarily long failed transaction, and then print several GB which we'd be forced to keep in memory. IMO we must have a limit to the amount we're buffering here.